[docs] Formatting and proper noun tune-up (#4200)
drud/ddev: docs/content/developers/building-contributing.md

Alternateively, you can see the latest build in action by visiting DDEV on [Gitpod](https://gitpod.io/#https://github.com/drud/ddev).

## Testing a PR

Each build of a PR has artifacts created in GitHub, so you can click the details of the [PR Build](https://github.com/drud/ddev/actions/workflows/pr-build.yml) test, choose the PR you want to work with, and download the artifacts you need.

After you download and unzip the appropriate binary, you can place it in your `$PATH`. The easiest way to do this if you’re using Homebrew is `brew unlink ddev` and then `unzip ddev.zip && chmod +x ddev && mv ddev /usr/local/bin/ddev`.

Verify the replacement worked via `ddev -v`. The output should be something like `ddev version v1.19.1-42-g5334d3c1`, instead of the regular `ddev version v1.19.1`.

(On macOS Big Sur these downloaded binaries are not signed, so you’ll want to `xattr -r -d com.apple.quarantine /path/to/ddev` in order to use them. The binaries on the master branch and the final binaries in any release *are* signed.)

You do not typically have to install anything else other than the downloaded binary; when you run it it will access any Docker images that it needs.

> After you’re done, you can remove the downloaded binary and `brew link ddev`.

Reviewer: Please don't use anything but real apostrophes. Other punctuation causes havoc with text editors and othere things.

> After you’re done, you can remove the downloaded binary and `brew link ddev`.

Author: Curly quotes [are more typographically correct](https://practicaltypography.com/straight-and-curly-quotes.html), which is why I called that out as an item to fix in the PR goals—since it’s specifically documentation in Markdown.

## Open in Gitpod

[Gitpod](https://www.gitpod.io) provides a quick, preconfigured DDEV experience in the browser for testing a PR easily without the need to set up an environment. In any PR you can use the URL `https://gitpod.io/#https://github.com/drud/ddev/pulls/<YOUR-PR>` to open that PR and build it in Gitpod.

To open and work on DDEV you can use the button below.
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/drud/ddev)

> If you want to run a web project, you can check it out into `/workspace/<yourproject>` and use it as usual. The things you’re familiar with work normally, except that `ddev-router` does not run.

Reviewer: ```suggestion
If you want to run a web project, you can check it out into `/workspace/<yourproject>` and use it as usual. The things you're familiar with work normally, except that `ddev-router` does not run.
```

Not even sure what you're using here. But use straight apostrophe.

`ddev version` should show you that you are using the correct webtag, and `ddev start` will show it.

It’s easiest to do this using Gitpod (see above) because Gitpod already has `docker buildx` all set up for you and the built DDEV binary is in the `$PATH`.

## Pull Requests and PR Preparation

> When preparing your pull request, please use a branch name like `2022_<your_username>_short_description` so it’s easy to identify you as the author.

Reviewer: ```suggestion
When preparing your pull request, please use a branch name like `2022MMDD_<your_username>_short_description` (like `20230901_rfay_short_description`) so it’s easy to identify you as the author.
```

## Docker Image Changes

If you make changes to a Docker image (like `ddev-webserver`), it won’t have any effect unless you:

> * Push an image with a specific tag by navigating to the image directory (like `containers/ddev-webserver`), and running `make container VERSION=<branchname>`.

Reviewer: ```suggestion
* Push an image with a specific tag by navigating to the image directory (like `containers/ddev-webserver`), and running `make push DOCKER_REPO=youruser/yourimage VERSION=<branchname>`. 
```

> * Multi-arch images require you to have a Buildx builder, so `docker buildx create --name ddev-builder-multi --use`.

Reviewer: ```suggestion
* Multi-arch images require you to have a buildx builder, so `docker buildx create --name ddev-builder-multi --use`.
```

Hmm, they don't use caps for buildx anywhere, so not sure it's reasonable to do so here. https://docs.docker.com/engine/reference/commandline/buildx/

Author: I don’t remember, but I probably took this as (admittedly sketchy) justification:
![Screen Shot 2022-09-25 at 11 02 29 AM@2x](https://user-images.githubusercontent.com/2488775/192158285-9aff2056-8f91-411c-8ef0-923cc977f232.png)

Author: Er, no, probably here: https://github.com/docker/buildx

They seem to use `buildx` in lowercase—often in backticks—describing the command and “Buildx” talking about the product. Still not totally consistent, though.

Author: Similar treatment here in marketing material: https://www.docker.com/blog/how-to-rapidly-build-multi-architecture-images-with-buildx/

Author: I’m going to double down on “Buildx” being the correct product name given the evidence!

3. All code should follow the guidelines covered in [Effective Go](http://golang.org/doc/effective_go.html) and [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments).
4. Comment the code. Tell us the why, the history and the context.
5. Document *all* declarations and methods, even private ones. Declare expectations, caveats and anything else that may be important. If a type gets exported, having the comments already there will ensure it’s ready.
6. Variable name length should be proportional to its context and no longer. `noCommaALongVariableNameLikeThisIsNotMoreClearWhenASimpleCommentWouldDo`. In practice, short methods will have short variable names and globals will have longer names.
7. No underscores in package names. If you need a compound name, step back, and re-examine why you need a compound name. If you still think you need a compound name, lose the underscore.
8. All tests should run with `go test` and outside tooling should not be required. No, we don’t need another unit testing framework. Assertion packages are acceptable if they provide *real* incremental value.

> 9. Even though we call these “rules” above, they are actually just guidelines. Since you’ve read all the rules, you now know that.

Reviewer: ```suggestion
9. Even though we call these "rules" above, they are actually just guidelines. Since you’ve read all the rules, you now know that.
```

Again, please use straight quotes and apostrophes everywhere. All editors know what to do with them.
